Mathlib/Computability/CFG.lean

```diff
@@ -0,0 +1,96 @@
+/-
+Copyright (c) 2023 Martin Dvorak. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Martin Dvorak
+-/
+import Mathlib.Computability.Language
+
+/-!
+# Context-Free Grammar
+
+This file contains the definition of a Context-Free Grammar (CFG), which is a grammar that has a
+single nonterminal symbol on the left-hand side of each rule.
+Derivation by a grammar is inherently nondeterministic.
+-/
+
+universe u
+
+/--
+The type of symbols is the disjoint union of terminals `T` and nonterminals `N`.
+We do not require `T` and `N` to be finite in this definition.
+As a result, we do not need to copy the typeclass instances `[Fintype T]` and `[Fintype N]`
+alongside our type parameters (which would appear in almost every lemma).
+Instead, later we work in terms of a list of rewrite rules, which is finite by definition and from
+which we could infer that only a finite set of terminals and a finite set of nonterminals can occur.
+-/
+inductive Symbol (T : Type u) (N : Type u)
```

Comment:
why do you want to use this instead of `sum`? Just to make the purpose of both sides more obvious?

Reply:
Yes.

Reply:
I got several positive reactions to this design decision. For such an important data type, names are better than generic `sum`.